tools/ecpunpack.cpp

```diff
@@ -0,0 +1,277 @@
+  {"LFE5U-85",  0x41113043, 13294, 1136, 0},
+  {"LFE5UM-25", 0x01111043, 7562,   592, 0},
+  {"LFE5UM-45", 0x01112043, 9470,   846, 2},
+  {"LFE5UM-85", 0x01113043, 13294, 1136, 0},
+};
+
+const uint8_t preamble[0x8] = {0xFF, 0xFF, 0xBD, 0xB3, 0xFF, 0xFF, 0xFF, 0xFF};
+
+class Bitstream {
+public:
+  Bitstream(uint8_t *_dat, int _len) : dat(_dat), len(_len) {};
+  Bitstream(std::string file) {
+    std::ifstream ifs(file, std::ios::binary|std::ios::ate);
+    std::ifstream::pos_type pos = ifs.tellg();
+    len = pos;
+    dat = new uint8_t[len];
+    ifs.seekg(0, std::ios::beg);
+    ifs.read((char *)dat, len);
+  }
+
+  uint8_t GetByte() {
+    assert(idx < len);
+    uint8_t val = dat[idx++];
+    UpdateCRC16(val);
+    return val;
+  }
+
+  void GetBytes(uint8_t *out, int count) {
+    for(int i = 0; i < count; i++) out[i] = GetByte();
+  }
+
+  void SkipBytes(int count) {
+    for(int i = 0; i < count; i++) GetByte();
+  }
+
+  uint32_t GetUint32() {
+    uint8_t tmp[4];
+    GetBytes(tmp, 4);
+    return (tmp[0] << 24UL) | (tmp[1] << 16UL) | (tmp[2] << 8UL) | (tmp[3]);
+  }
+
+  void UpdateCRC16(uint8_t val) {
+    int bit_flag;
+    for(int i = 7; i >= 0; i--) {
+      bit_flag = crc16 >> 15;
+
+      /* Get next bit: */
+      crc16 <<= 1;
+      crc16 |= (val >> i) & 1; // item a) work from the least significant bits
+
+      /* Cycle check: */
+      if(bit_flag)
+          crc16 ^= CRC16_POLY;
+    }
+
+  }
+
+  bool FindPreamble(const uint8_t *preamble, int pre_len) {
+    for(int i = idx; i <= (len - pre_len); i++) {
+      if(memcmp(preamble, dat+i, pre_len) == 0) {
+        std::cerr << "found preamble at offset " << std::dec << i << std::endl;
+        idx = i + pre_len;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  uint16_t GetCRC16() {
+    // item b) "push out" the last 16 bits
+    int i, bit_flag;
+    for (i = 0; i < 16; ++i) {
+        bit_flag = crc16 >> 15;
+        crc16 <<= 1;
+        if(bit_flag)
+            crc16 ^= CRC16_POLY;
+    }
+
+    return crc16;
+  }
+  void ResetCRC16() { crc16 = 0; }
+  bool CheckCRC16() {
+      /*TODO*/
+    uint16_t act_crc = GetCRC16();
+    uint8_t crc_bytes[2];
+    GetBytes(crc_bytes, 2);
+    uint16_t exp_crc = (crc_bytes[0] << 8) | crc_bytes[1];
+    if(act_crc != exp_crc)
+      std::cerr << "crc fail, calculated " << std::hex << act_crc << ", expected " << exp_crc << std::endl;
+    ResetCRC16();
+    return exp_crc == act_crc;
+  }
+
+  bool Done() {
+    return (idx >= len);
+  }
+private:
+  uint16_t crc16 = 0x0;
+  uint8_t *dat;
+  int len;
+  int idx = 0;
+};
+
+struct ChipConfig {
+
+  ChipConfig(DeviceInfo _dev) : dev(_dev) {
+    chip.resize(dev.n_frames);
+    for(int i = 0; i < dev.n_frames; i++)
+      chip[i].resize(dev.frame_data_bits, false);
+  };
+
+  void LoadFrame(int frame, Bitstream &bs) {
+    int frame_bytes = (dev.frame_data_bits + dev.frame_pad_bits) / 8;
+    uint8_t *data = new uint8_t[frame_bytes];
```

Comment:
I’d like to see the standard smart pointers used rather than raw pointers.  In this case std::unique_ptr should be used for *data.  Smart pointers will keep us from having to worry about memory management (for the most part) and let us focus more on the domain code as the codebase gets more complex. http://en.cppreference.com/w/cpp/memory/unique_ptr

Reply:
Since this is an array, maybe it should be std:array